File: nest-cloud/src/data/mappers/ProductMapper.ts
import DomainMapper from "src/domain/mappers/DomainMapper";
import { Product } from "src/domain/model/Product";
import { ProductEntity } from "../entity/ProductEntity";

export class ProductMapper implements DomainMapper<Product, ProductEntity>{
    toDomain(entity: ProductEntity): Product {
        return new Product(
            entity.id,
            entity.name,
            entity.description,
            entity.product_code,
            entity.form,
            entity.material,
            entity.unit,
            entity.quantity_per_unit,
            entity.price,
            entity.unit_price,
            entity.sub_category_id,
            []
        )
    }

    fromDomain(domain: Product): ProductEntity {
        return new ProductEntity(
            domain.productId,
            domain.name,
            domain.description,
            domain.productCode,
            domain.form,
            domain.material,
            domain.unit,
            domain.quantityPerUnit,
            domain.price,
            domain.unitPrice,
            domain.subCategoryId,
            null,
            null
        )
    }

}

File: android-app/data/src/main/java/com/starlight/module/data/datasource/mappers/ProductMapper.kt
package com.starlight.module.data.datasource.mappers

import com.starlight.module.data.datasource.entity.ProductEntity
import com.starlight.module.domain.mappers.DomainMapper
import com.starlight.module.domain.model.Product


class ProductMapper : DomainMapper<ProductEntity, Product> {
    override fun toDomain(entity: ProductEntity): Product {
        return Product(
            entity.productId,
            entity.name,
            entity.description,
            entity.productCode,
            entity.form,
            entity.material,
            entity.unit,
            entity.quantityPerUnit,
            entity.price,
            entity.unitPrice,
            entity.subCategoryId,
            false,
        )
    }

    fun toDomain(entity: ProductEntity, isFavorite: Boolean): Product {
        val product = toDomain(entity)
        product.isFavorite = isFavorite
        return product
    }

    override fun fromDomain(model: Product): ProductEntity {
        return ProductEntity(
            model.id,
            model.name,
            model.description,
            model.productCode,
            model.form,
            model.material,
            model.unit,
            model.quantityPerUnit,
            model.price,
            model.unitPrice,
            model.subCategoryId
        )
    }
}